uds/src/lib.rs

```diff
@@ -0,0 +1,263 @@
+#[cfg(test)]
+extern crate tokio_current_thread;
+#[cfg(test)]
+extern crate tokio_io;
+
+use futures::future::{self, Future, FutureResult};
+use futures::stream::Stream;
+use multiaddr::{AddrComponent, Multiaddr};
+use std::io::Error as IoError;
+use std::path::PathBuf;
+use libp2p_core::Transport;
+use tokio_uds::{UnixListener, UnixStream};
+
+/// Represents the configuration for a Unix domain sockets transport capability for libp2p.
+///
+/// The Unixs sockets created by libp2p will need to be progressed by running the futures and
+/// streams obtained by libp2p through the tokio reactor.
+#[derive(Debug, Clone)]
+pub struct UdsConfig {
+}
+
+impl UdsConfig {
+    /// Creates a new configuration object for TCP/IP.
+    #[inline]
+    pub fn new() -> UdsConfig {
+        UdsConfig {}
+    }
+}
+
+#[cfg(unix)]
+impl Transport for UdsConfig {
+    type Output = UnixStream;
+    type Listener = Box<Stream<Item = Self::ListenerUpgrade, Error = IoError>>;
+    type ListenerUpgrade = FutureResult<(Self::Output, Self::MultiaddrFuture), IoError>;
+    type MultiaddrFuture = FutureResult<Multiaddr, IoError>;
+    type Dial = Box<Future<Item = (UnixStream, Self::MultiaddrFuture), Error = IoError>>;
+
+    fn listen_on(self, addr: Multiaddr) -> Result<(Self::Listener, Multiaddr), (Self, Multiaddr)> {
+        if let Ok(path) = multiaddr_to_path(&addr) {
+            let listener = UnixListener::bind(&path);
+            // We need to build the `Multiaddr` to return from this function. If an error happened,
+            // just return the original multiaddr.
+            match listener {
+                Ok(_) => {},
+                Err(_) => return Err((self, addr)),
+            };
+
+            debug!("Now listening on {}", addr);
+            let new_addr = addr.clone();
+
+            let future = future::result(listener)
+                .map(move |listener| {
+                    // Pull out a stream of sockets for incoming connections
+                    listener.incoming().map(move |sock| {
+                        debug!("Incoming connection on {}", addr);
+                        future::ok((sock, future::ok(addr.clone())))
+                    })
+                })
+                .flatten_stream();
+            Ok((Box::new(future), new_addr))
+        } else {
+            Err((self, addr))
+        }
+    }
+
+    fn dial(self, addr: Multiaddr) -> Result<Self::Dial, (Self, Multiaddr)> {
+        if let Ok(path) = multiaddr_to_path(&addr) {
+            debug!("Dialing {}", addr);
+            let fut = UnixStream::connect(&path).map(|t| (t, future::ok(addr)));
+            Ok(Box::new(fut) as Box<_>)
+        } else {
+            Err((self, addr))
+        }
+    }
+
+    fn nat_traversal(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        if server == observed {
+            Some(observed.clone())
+        } else {
+            None
+        }
+    }
+}
+
+#[cfg(not(unix))]
+impl Transport for UdsConfig {
```

Comment:
How about only defining `UdsConfig` and its `Transport` impl for Unix and not implementing the transport for other systems at all?

Reply:
I'm kind of hesitant. Right now creating the `WsConfig` is different whether you use it with emscripten or not, but other than that there's nothing platform-specific in libp2p.